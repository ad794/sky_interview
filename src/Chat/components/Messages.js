import React, { Component } from 'react';
import { View, FlatList } from 'react-native';
import PropTypes from 'prop-types';
import Message from './Message';
import Avatar from './Avatar';
import styles from './styles';

class Messages extends Component {
  static propTypes = {
    data: PropTypes.array
  };

  renderFlatList = () => {
    const { data, members } = this.props;
    return (
      <FlatList
        data={data}
        extraData={members}
        renderItem={({ item }) => this.renderItem(item)}
        keyExtractor={item => item.id}
      />
    );
  };

  renderItem = ({ userId, message, timestamp }) => {
    const member = this.props.members[userId];
    return (
      <View style={styles.card}>
        <View style={{ width: 100, height: 100 }}>{member && <Avatar uri={member.avatar} />}</View>
        <Message message={message} timestamp={timestamp} member={member} />
      </View>
    );
  };
  render() {
    return this.renderFlatList();
  }
}

export default Messages;
