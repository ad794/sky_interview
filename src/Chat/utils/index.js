export const timestampToDate = (timestamp) => {
  const date = new Date(timestamp * 1000);
  return `${date.toLocaleDateString()} at ${formatTime(date)}`;
};

formatTime = (date_obj) => {
  // formats a javascript Date object into a 12h AM/PM time string
  let hour = date_obj.getHours();
  let minute = date_obj.getMinutes();
  const amPM = hour > 11 ? 'PM' : 'AM';
  if (hour > 12) {
    hour -= 12;
  } else if (hour == 0) {
    hour = '12';
  }
  if (minute < 10) {
    minute = `0${minute}`;
  }
  return `${hour}:${minute}${amPM}`;
};
export const findById = function (tree, testID) {
  if (tree.props && tree.props.testID === testID) {
    return tree;
  }
  if (tree.children && tree.children.length > 0) {
    const childs = tree.children;
    for (let i = 0; i < childs.length; i++) {
      const item = findById(childs[i], testID);
      if (typeof item !== 'undefined') {
        return item;
      }
    }
  }
};
