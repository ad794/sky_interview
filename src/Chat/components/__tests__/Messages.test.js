import React from 'react';
import renderer from 'react-test-renderer';
import Messages from '../Messages';

import { findById } from '../../utils';

const mock = {
  data: [
    {
      id: 'b03569ae-ccbf-4975-8040-4daba638b405',
      userid: '1',
      message: 'message 1'
    },
    {
      id: 'b03569ae-ccbf-4975-8040-4daba638b407',
      userId: '2',
      message: 'message 1'
    }
  ],
  members: [
    {
      1: {
        userId: '1',
        email: 'test@test.com',
        avatar: ''
      }
    },
    {
      2: {
        userId: '2',
        email: 'test@test.com',
        avatar: ''
      }
    }
  ]
};

describe('messages component', () => {
  const messagesComponentTree = renderer.create(<Messages {...mock} />).toJSON();
  const messagesComponent = renderer.create(<Messages {...mock} />).getInstance();
  it('renders correctly', () => {
    expect(messagesComponentTree).toMatchSnapshot();
  });
  it('renders with correct props', () => {
    expect(messagesComponent.props).toEqual(mock);
  });
});
