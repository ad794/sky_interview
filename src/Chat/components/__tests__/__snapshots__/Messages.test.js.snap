// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messages component renders correctly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "id": "b03569ae-ccbf-4975-8040-4daba638b405",
        "message": "message 1",
        "userid": "1",
      },
      Object {
        "id": "b03569ae-ccbf-4975-8040-4daba638b407",
        "message": "message 1",
        "userId": "2",
      },
    ]
  }
  disableVirtualization={false}
  extraData={
    Array [
      Object {
        "1": Object {
          "avatar": "",
          "email": "test@test.com",
          "userId": "1",
        },
      },
      Object {
        "2": Object {
          "avatar": "",
          "email": "test@test.com",
          "userId": "2",
        },
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "backgroundColor": "#F5FCFF",
            "flex": 1,
            "flexDirection": "row",
            "marginBottom": 15,
          }
        }
      >
        <View
          style={
            Object {
              "height": 100,
              "width": 100,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
                "marginLeft": 10,
              }
            }
          >
            <View
              accessible={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              testID="showEmail"
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                    "marginLeft": 10,
                  }
                }
                testID="messageText"
              >
                message 1
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "flex": 1,
                "justifyContent": "flex-end",
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              1970-1-1 at 12:00AM
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "backgroundColor": "#F5FCFF",
            "flex": 1,
            "flexDirection": "row",
            "marginBottom": 15,
          }
        }
      >
        <View
          style={
            Object {
              "height": 100,
              "width": 100,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
                "marginLeft": 10,
              }
            }
          >
            <View
              accessible={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              testID="showEmail"
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                    "marginLeft": 10,
                  }
                }
                testID="messageText"
              >
                message 1
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "flex": 1,
                "justifyContent": "flex-end",
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              1970-1-1 at 12:00AM
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
